02-peer-protocol.md

```diff
@@ -50,10 +67,451 @@ pubkey corresponding to the funding output nothing prevents duplicative channel
+from either the index `0` or `1`.
+
+Given an index `i`, `J` is to be found as follows:
+  - For `c` in the range 0 to 255,
+    - Find `sha = sha256( G || i || c)` where `G` is the 33-byte
+      compressed point representation of the secp256k1 generator point
+    - Create a positive compressed point representation from `sha` as
+      `02 || sha`
+    - Check if the resulting point is on the secp256k1 curve. If so,
+      this is the `J` point for this index. If not, continue with the next
+      value for `c`
+
+This algorithm will give the following `J` points:
+
+| `i`, index | J                                                                |
+|------------|------------------------------------------------------------------|
+| 0          |0296f47ec8e6d6a9c3379c2ce983a6752bcfa88d46f2a6ffe0dd12c9ae76d01a1f|
+| 1          |023f9976b86d3f1426638da600348d96dc1f1eb0bd5614cc50db9e9a067c0464a2|
+
+##### Requirements
+The sender:
+  - MUST calculate the `e`, `sig`, and `p2` for the key-pair that the
+    corresponding utxo is locked to
+
+The receiver:
+  - if the sha256(P2) is NOT equal to the `podle_h2` communicated during
+      the initiation phase:
+    - MAY error
+    - MUST wait until a `tx_add_input` with a matching `podle_h2` is received
+      and verified before sending their own `tx_add_input`
+    - SHOULD abort the verification process
+  - if the utxo `prevtx_scriptpubkey` is NOT locked to by the provided `pubkey`. (This may
+      require computing the hash160 of the pubkey if the script type is P2WPKH):
+    - MUST error
+  - if the signature is incorrect, or `sha256(kG||kJ||P||P2||node_id) != e`:
+    - MUST error
+  - if the sender errors the transaction collaboration after `tx_complete` has been
+    negotiated:
+     - SHOULD broadcast a `blacklist_podle` gossip message for the `podle_h2` of this
+       transaction collaboration
+
+##### Rationale
+The PoDLE protocol allows a peer to prove ownership of a utxo involved in a
+transaction collaboration. This serves as a rate-limiting mechanism for
+malicious behavior and allows honest nodes to share information about
+malicious attempts. A malicious node attempting to probe a `contributor`'s
+utxo shot, in the best case, is only granted one attempt per node. We limit
+probe attempts across the network by gossiping bad actor's H2 commitments.
+
+PoDLEs will not stop a determined attacker from repeatedly probing a node,
+but it does increase the cost of doing so and effectively rate-limits
+such attempts.
+
+
+### The `tx_add_output` Message
+
+This message contains information for outputs to a funding transaction.
+
+1. type: 67 (`tx_add_output`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`u16`:`serial_id`]
+    * [`u64`:`sats`]
+    * [`u16`:`scriptlen`]
+    * [`scriptlen*byte`:`script`]
+
+#### Requirements
+
+Either node:
+  - MAY omit this message
+
+The sending node:
+  - MUST add all sent outputs to the collaborative transaction
```

Comment:
Isn't there only one? Same question applies to the other three sent/received input/output places where it says "all"